src/qt/ScopeGuard.h

```diff
@@ -0,0 +1,77 @@
+#ifndef MONERO_GUI_SCOPEGUARD_H
```

Comment:
Now that xiphon mentioned it, might be a good idea to add following to the top:

```
Copyright (C) 2018 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com, author Sérgio Martins <sergio.martins@kdab.com>
Contact: https://www.qt.io/licensing/
```

Reply:
I would say you should copy the whole license header from the original Qt code.

Reply:
Heh this becomes a problem now. The license is https://www.gnu.org/licenses/lgpl-3.0.html
Is it compatible with Monero-GUI?

Reply:
```
/****************************************************************************
**
** Copyright (C) 2018 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com, author Sérgio Martins <sergio.martins@kdab.com>
** Contact: https://www.qt.io/licensing/
**
** This file is part of the QtCore module of the Qt Toolkit.
**
** $QT_BEGIN_LICENSE:LGPL$
** Commercial License Usage
** Licensees holding valid commercial Qt licenses may use this file in
** accordance with the commercial license agreement provided with the
** Software or, alternatively, in accordance with the terms contained in
** a written agreement between you and The Qt Company. For licensing terms
** and conditions see https://www.qt.io/terms-conditions. For further
** information use the contact form at https://www.qt.io/contact-us.
**
** GNU Lesser General Public License Usage
** Alternatively, this file may be used under the terms of the GNU Lesser
** General Public License version 3 as published by the Free Software
** Foundation and appearing in the file LICENSE.LGPL3 included in the
** packaging of this file. Please review the following information to
** ensure the GNU Lesser General Public License version 3 requirements
** will be met: https://www.gnu.org/licenses/lgpl-3.0.html.
**
** GNU General Public License Usage
** Alternatively, this file may be used under the terms of the GNU
** General Public License version 2.0 or (at your option) the GNU General
** Public license version 3 or any later version approved by the KDE Free
** Qt Foundation. The licenses are as published by the Free Software
** Foundation and appearing in the file LICENSE.GPL2 and LICENSE.GPL3
** included in the packaging of this file. Please review the following
** information to ensure the GNU General Public License requirements will
** be met: https://www.gnu.org/licenses/gpl-2.0.html and
** https://www.gnu.org/licenses/gpl-3.0.html.
**
** $QT_END_LICENSE$
**
****************************************************************************/
```

appears to be the full license

https://github.com/RSATom/Qt/blob/master/qtbase/src/corelib/tools/qscopeguard.h

Reply:
looks ok?

https://opensource.stackexchange.com/questions/6405/lgpl-licensed-code-reuse-in-mit-licensed-project

Reply:
This seems problematic at this point. We like to have something like scope guard to have RAII resource management, but without upgrading the Qt I think we cannot use this source without forcing the whole Monero-GUI to be LGPLv3, am I right?

Do we need to reimplement scope guard by ourselves? And if it is the case, it probably cannot be called `qScopeGuard` anymore as it won't have the same properties. 
Then we need to have something ours, which fallbacks to `qScopeGuard` if it is present. Am I right?

Reply:
> Heh this becomes a problem now. The license is https://www.gnu.org/licenses/lgpl-3.0.html
> Is it compatible with Monero-GUI?

At the first glance, we ain't doing anything malicious, nor we violate the license.

We don't claim the copyright on the code and we copy the original license and copyright.

Reply:
There is also this: https://spdx.org/licenses/Qt-LGPL-exception-1.1.html

Reply:
I am removing this scope guard and using https://github.com/ricab/scope_guard

Reply:
Any ideas where to put the scope guard related files? it does not seem right to put it in `src/qt`, should I create `src/misc` ?

Reply:
FWIW we already have a Qt file in the source code tree (MoneroSettings) so if we violate the license we did for a while.

Reply:
> I am removing this scope guard and using https://github.com/ricab/scope_guard

Great, that is a good one.

>Any ideas where to put the scope guard related files? it does not seem right to put it in `src/qt`, should I create `src/misc` ?

That's correct. But we already have some unrelated things in `src/qt`, please place it there for now.

Reply:
Hmm I still think that having LGPL sources in the app is problematic. This LGPL has to be stated somewhere in the app (about), the app is distributed also in the binary forms. I am not sure whether this is safe. LGPL states 

> A program that contains no derivative of any portion of the Library, but is designed to work with the Library by being compiled or linked with it, is called a "work that uses the Library". Such a work, in isolation, is not a derivative work of the Library, and therefore falls outside the scope of this License.

I think that means that whole Monero-GUI would have to be LGPL when we use the source file directly. https://www.quora.com/If-my-code-uses-some-code-licensed-under-LGPL-can-I-release-the-whole-software-under-MIT

Reply:
So what do you think guys? I would strongly prefer to use Unlicensed https://github.com/ricab/scope_guard to avoid such license pains. Do you agree?

If yes, should I create `src/misc` or `src/utils` where to put these new files or should I use some existing dir?

Reply:
> I would strongly prefer to use Unlicensed https://github.com/ricab/scope_guard to avoid such license pains. Do you agree?

Yes.

Already replied to this, see https://github.com/monero-project/monero-gui/pull/2829#discussion_r422211422

Reply:
> Already replied to this, see [#2829 (comment)](https://github.com/monero-project/monero-gui/pull/2829#discussion_r422211422)

Ah I see, GitHub didn't refresh the conversation.

Reply:
ok done, pls check